source/rpt/rpt_cell_reconstruction.cc

```diff
@@ -0,0 +1,626 @@
+  int                                 best_cell;
+
+  // Use least squared error cell by cell for all counts per detector and
+  // all vertices of the cell
+  for (const auto &cell : triangulation.cell_iterators_on_level(level))
+    {
+      bool is_candidate = false;
+
+      // Check if candidate cell
+      for (auto &id : candidate)
+        {
+          if (id == cell->index())
+            is_candidate = true;
+        }
+      if (is_candidate)
+        {
+          error = 0;
+          for (unsigned int j = 0; j < detectors.size(); j++)
+            {
+              for (unsigned int i = 0; i < cell->n_vertices(); i++)
+                {
+                  unsigned int vertex_index = cell->vertex_index(i);
+                  error +=
+                    std::pow(std::fabs(
+                               particle_reconstruction_counts[j] -
+                               map_vertices_index[vertex_index].second[j]),
+                             2);
+                }
+            }
+          cellid_error.push_back(std::make_pair(cell->index(), error));
+        }
+    }
+
+  // Sorted index of cell by error
+  std::sort(cellid_error.begin(),
+            cellid_error.end(),
+            [](std::pair<int, double> &pair1, std::pair<int, double> &pair2) {
+              return (pair1.second < pair2.second);
+            });
+
+  // Return the best cell index
+  return best_cell = cellid_error[0].first;
+}
+
+template <int dim>
+void
+RPTCellReconstruction<dim>::read_counts()
+{
+  // Read counts for reconstruction
+  const std::string filename =
+    parameters.reconstruction_param.reconstruction_counts_file;
+  std::ifstream counts_file(filename);
+
+  std::string skip;
+  std::getline(counts_file, skip); // Skip header line
+  std::vector<double> values;
+  std::copy(std::istream_iterator<double>(counts_file),
+            std::istream_iterator<double>(),
+            std::back_inserter(values));
+
+  // Get the number of detector
+  unsigned int number_of_detectors = detectors.size();
+
+  // Extract positions, create point objects and detectors
+  for (unsigned int i = 0; i < values.size(); i += number_of_detectors)
+    {
+      std::vector<double>::const_iterator first = values.begin() + i;
+      std::vector<double>::const_iterator last =
+        values.begin() + i + number_of_detectors;
+
+      std::vector<double> particle_counts(first, last);
+
+      reconstruction_counts.push_back(particle_counts);
+    }
+}
+
+template <int dim>
+void
+RPTCellReconstruction<dim>::export_positions()
+{
+  std::ofstream myfile;
+  std::string   sep;
+  std::string   filename =
+    parameters.reconstruction_param.reconstruction_positions_file;
+  myfile.open(filename);
+
+  // Headers with space or comma
+  if (filename.substr(filename.find_last_of(".") + 1) == ".dat")
+    {
+      myfile
+        << "unknown_positions_id volumes particle_positions_x particle_positions_y particle_positions_z cell_level";
```

Comment:
What does unknown_position mean here? The position reconstructed from the detectors?